spec/std/yaml/from_yaml_spec.cr

```diff
@@ -0,0 +1,374 @@
+      values.each do |value|
+        expect_raises(YAML::ParseException) do
+          Bool.from_yaml %("#{value}")
+        end
+      end
+    end
+
+    it "should raise if falsey is quoted with single quotes" do
+      values.each do |value|
+        expect_raises(YAML::ParseException) do
+          Bool.from_yaml "'#{value}'"
+        end
+      end
+    end
+  end
+
+  it "should raise if not a bool value" do
+    expect_raises(YAML::ParseException) do
+      Bool.from_yaml "hello"
+    end
+  end
+end
+
+[Int8, Int16, Int32, Int64, UInt8, UInt16, UInt32, UInt64].each do |type|
+  describe "#{type.name}.from_yaml" do
+    it "should serialize and deserialize" do
+      type.from_yaml(type.new("1").to_yaml).should eq type.new("1")
+    end
+
+    it "should parse a number into an #{type.name}" do
+      type.from_yaml("1").should eq type.new("1")
+    end
+
+    it "should raise if quoted with double quotes" do
+      expect_raises(YAML::ParseException) do
+        type.from_yaml %("1")
+      end
+    end
+
+    it "should raise if quoted with single quotes" do
+      expect_raises(YAML::ParseException) do
+        type.from_yaml "'1'"
+      end
+    end
+
+    it "should raise when not a number" do
+      expect_raises(YAML::ParseException) do
+        type.from_yaml("hi")
+      end
+    end
+  end
+end
+
+[Float32, Float64].each do |type|
+  describe "#{type.name}.from_yaml" do
+    it "should serialize and deserialize" do
+      type.from_yaml(type.new("1.0").to_yaml).should eq type.new("1.0")
+    end
+
+    it "should parse a number into an #{type.name}" do
+      type.from_yaml("1").should eq type.new("1")
+    end
+
+    it "should parse a float into an #{type.name}" do
+      type.from_yaml("1.1").should eq type.new("1.1")
+    end
+
+    it "should raise if quoted with double quotes" do
+      expect_raises(YAML::ParseException) do
+        type.from_yaml %("1.1")
+      end
+    end
+
+    it "should raise if quoted with single quotes" do
+      expect_raises(YAML::ParseException) do
+        type.from_yaml "'1.1'"
+      end
+    end
+
+    it "should raise when not a float or number" do
+      expect_raises(YAML::ParseException) do
+        type.from_yaml("hi")
+      end
+    end
+  end
+end
+
+describe "String.from_yaml" do
+  it "should serialize and deserialize" do
+    String.from_yaml("hi".to_yaml).should eq "hi"
+  end
+
+  it "should parse a string" do
+    String.from_yaml("hello").should eq "hello"
+  end
+
+  it "should parse a single quoted string" do
+    String.from_yaml("'hello'").should eq "hello"
+  end
+
+  it "should parse a double quoted string" do
+    String.from_yaml(%("hello")).should eq "hello"
+  end
+
+  it "should parse a literal string" do
+    String.from_yaml("|\n  hello").should eq "hello"
+  end
+
+  it "should parse a folded string" do
+    String.from_yaml(">\n  hello").should eq "hello"
+  end
+
+  context "reserved values" do
```

Comment:
I'm thinking about splitting these reserved values up and cover each set in the context of the specific type. It's not an inherent responsibility of the `string` type to recognize them. Only when a specific type is supported, it's special values 
or formats are not considered strings but values of that type. `String` is only affected because it's the default type. If support for additional types is added, that would not mean to modify existing specs but just add the additional behaviour in a cohesive set of specs which cover that special values are not recognized as strings. In this case it actually makes sense to differentiate between individual specs for each type of reserved values.
The specs for raising if a string is to be read from a reserved value and passing if quoted could easily made reusable to employ them at these individual locations. This way, we don't need duplicate value literals as well, because they already exists at the designated places.

Reply:
I don't think that "1.0" is a valid yaml string, unless quoted. So why should it not be string's responsibility to detect that?

Reply:
It is only an invalid string because it is a float value format. If the parser schema wouldn't include float type (like the Failsafe Schema), `1.0` would be recognized as a string: `a sequence of zero or more Unicode characters.`

Reply:
I don't see anywhere in the spec where it makes clear how the parser should handle situations like this. Maybe it is a string in the failsafe schema but we don't have an option to select schema do we? If we're not in the fail safe scheme then surely reading a string which is a valid float, which is unquoted, is incorrect. The way that yaml is parsed should be entirely independent of the way that its converted into crystal types. A certain value shouldn't be parsed one way of were expecting a string and the other way if we're expecting a float. Doing so is leaking details of the parsing. The fact is that 99% of yaml parsers will parse an unquoted 1.0 as a number. If we don't do the same, then our parser is incorrect. 

That leads me to the conclusion that the way we're doing parsing is incorrect. We should be parsing tokens and deciding what type they are inside a single next method. We shouldn't be attempting to read a string and erroring if we get something which could be a float, we should have a single next method which returns the union of all yaml types, and parses the scalar to a concrete type irrespective of what type we've told the pullparser we're expecting. This is the same way it's done in the json pullparser.

Reply:
@RX14 I think these general considerations shouldn't go into a line comment. Will get lost easily this way...

Reply:
So do we need to make it so that String's always parse and not check reserved scalars. According to the spec...

```
This tag accepts all scalar values. It is is typically used as the fall-back tag for scalar nodes.
```

But to @RX14's point, all other languages will treat the values as float's/etc when read.

Reply:
It all depends on the used schema. If a parser supports Core Schema, it will of course recognize a float value. If the schema has no float, `1.0` is recognized as a string.

Reply:
Right, but we do recognize float/core schema.

Reply:
From a logical perspective, it's not a property of the string parser to recognize `1.0` as a reserved value but it is causatively tied to the float parser. That's why I think it's consequential to align the specs there.

Reply:
This way it's easier to track for example further additions to the parser which might come in the future because these, too, would be introduced in one piece of matching values and non-matching string values.

Reply:
What i'm suggesting is to move all the scalar parsing logic into `read_next`. `read_next` will read a scalar or whatever using libyaml, but then it will parse the value (if it's a scalar) to work out what scalar type it is. This makes the rest of the class simpler because the `read_string` method doesn't need to work out if it's a string: it's just called `read_next` which has set which scalar type the token is. This reduces duplication and is much more in-line with how the json pullparser works.